src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs

```diff
@@ -100,12 +106,44 @@ public async override Task ProcessAsync(TagHelperContext context, TagHelperOutpu
                 throw new ArgumentException(Resources.FormatAttributeIsRequired(RenderModeName, TagHelperName), nameof(RenderMode));
             }
 
-            var componentRenderer = ViewContext.HttpContext.RequestServices.GetRequiredService<IComponentRenderer>();
+            var requestServices = ViewContext.HttpContext.RequestServices;
+            var componentRenderer = requestServices.GetRequiredService<IComponentRenderer>();
+            var htmlRenderer = requestServices.GetRequiredService<HtmlRenderer>();
             var result = await componentRenderer.RenderComponentAsync(ViewContext, ComponentType, RenderMode, _parameters);
 
+            if (ViewContext.GetAutomaticComponentPersistencePreference() && (RenderMode == RenderMode.ServerPrerendered || RenderMode == RenderMode.WebAssemblyPrerendered))
```

Comment:
I'm not 100% sold on the idea of automatic persistence. There's the whole "principle of least surprise" thing and it is quite surprising that adding a second root component would - by default - deactivate the persistence of the first root component.

It would be less surprising if persistence was on if (and only if) there's a `<persist-component-state>`. That also makes it clear that you're emitting the state to the HTML. It could even be named more clearly to indicate that, e.g., `<emit-paused-component-state />`.

Reply:
I was trying to be a bit smart about it and optimize for the case where there is only one component on the page. If we are ok, we can just require you to have `save-persisted-state` on the page at all times.

The goal was to make it work automagically for 1 component and force you to be explicit if you were rendering more than one component into the page.